src/kdc/do_tgs_req.c

```diff
@@ -917,38 +896,52 @@ prepare_error_tgs (struct kdc_request_state *state,
 /* KDC options that require a second ticket */
 #define STKT_OPTIONS (KDC_OPT_CNAME_IN_ADDL_TKT | KDC_OPT_ENC_TKT_IN_SKEY)
 /*
- * Get the key for the second ticket, if any, and decrypt it.
+ * If req is a second-ticket request and a second ticket is present, decrypt
+ * it.  Set *stkt_out to an alias to the ticket with populated enc_part2.  Set
+ * *server_out to the server DB entry and *key_out to the ticket decryption
+ * key.
  */
 static krb5_error_code
 decrypt_2ndtkt(kdc_realm_t *kdc_active_realm, krb5_kdc_req *req,
-               krb5_flags flags, krb5_db_entry **server_out,
-               krb5_keyblock **key_out, const char **status)
+               krb5_flags flags, const krb5_ticket **stkt_out,
+               krb5_db_entry **server_out, krb5_keyblock **key_out,
+               const char **status)
 {
     krb5_error_code retval;
     krb5_db_entry *server = NULL;
+    krb5_keyblock *key = NULL;
     krb5_kvno kvno;
     krb5_ticket *stkt;
 
-    if (!(req->kdc_options & STKT_OPTIONS))
+    *stkt_out = NULL;
+    *server_out = NULL;
+    *key_out = NULL;
+
+    if (!(req->kdc_options & STKT_OPTIONS) || req->second_ticket == NULL ||
```

Comment:
btw, should we have a check that if `!(req->kdc_options & STKT_OPTIONS)` then `req->second_ticket == NULL` ?

Reply:
An interesting question.  Since someone could invent a new kdc-options flag requiring a second ticket, I think this check would only make sense if we also wanted to fail on unknown KDC options, and that would in turn only be desirable if the other major KDC implementations did so.  (RFC 4120 doesn't say what to do with unrecognized options.)